test/utils/Checkpoints.test.js

```diff
@@ -4,71 +4,144 @@ const { expect } = require('chai');
-      expect(await this.checkpoint.getAtBlock(this.tx1.receipt.blockNumber)).to.be.bignumber.equal('1');
-      expect(await this.checkpoint.getAtBlock(this.tx2.receipt.blockNumber)).to.be.bignumber.equal('2');
-      // Block with no new checkpoints
-      expect(await this.checkpoint.getAtBlock(this.tx2.receipt.blockNumber + 1)).to.be.bignumber.equal('2');
-      expect(await this.checkpoint.getAtBlock(this.tx3.receipt.blockNumber)).to.be.bignumber.equal('3');
-      expect(await this.checkpoint.getAtBlock(this.tx3.receipt.blockNumber + 1)).to.be.bignumber.equal('3');
-    });
+        it('lookup returns 0', async function () {
+          expect(await this.contract.lowerLookup(0)).to.be.bignumber.equal('0');
+          expect(await this.contract.upperLookup(0)).to.be.bignumber.equal('0');
+          expect(await this.contract.upperLookupRecent(0)).to.be.bignumber.equal('0');
+        });
+      });
 
-    it('reverts if block number >= current block', async function () {
-      await expectRevert(
-        this.checkpoint.getAtBlock(await web3.eth.getBlockNumber()),
-        'Checkpoints: block not yet mined',
-      );
+      describe('with checkpoints', function () {
+        beforeEach('pushing checkpoints', async function () {
+          this.checkpoints = [
+            { key: 2, value: '2' },
+            { key: 3, value: '3' },
+            { key: 5, value: '5' },
+            { key: 7, value: '8' },
+            { key: 11, value: '13' },
+          ];
+          for (const { key, value } of this.checkpoints) {
+            await this.contract.push(key, value);
+          }
+        });
 
-      await expectRevert(
-        this.checkpoint.getAtBlock(await web3.eth.getBlockNumber() + 1),
-        'Checkpoints: block not yet mined',
-      );
-    });
+        it('returns latest value', async function () {
+          expect(await this.contract.latest())
+            .to.be.bignumber.equal(last(...this.checkpoints).value);
+        });
+
+        it('cannot push values in the past', async function () {
+          await expectRevert(this.contract.push(last(...this.checkpoints).key - 1, '0'), 'Checkpoint: invalid key');
+        });
+
+        it('can update last value', async function () {
+          const newValue = '42';
 
-    it('multiple checkpoints in the same block', async function () {
-      const lengthBefore = await this.checkpoint.length();
-      await batchInBlock([
-        () => this.checkpoint.push(8, { gas: 100000 }),
-        () => this.checkpoint.push(9, { gas: 100000 }),
-        () => this.checkpoint.push(10, { gas: 100000 }),
-      ]);
-      const lengthAfter = await this.checkpoint.length();
-
-      expect(lengthAfter.toNumber()).to.be.equal(lengthBefore.toNumber() + 1);
-      expect(await this.checkpoint.latest()).to.be.bignumber.equal('10');
+          await this.contract.push(last(...this.checkpoints).key, newValue);
+          expect(await this.contract.latest()).to.be.bignumber.equal(newValue);
```

Comment:
Can we also check that this was not due to adding a new checkpoint? I.e. the number of checkpoints remained the same. Currently the test wouldn't detect if a new checkpoint was pushed.